File: deconv_comparison/binio.m
function H = binio()
% H = binio()
%
% Flat float64 binary exchange with Python. Column-major, with a sidecar
% '<name>.shape' text file. H.save(name,A) / H.load(name).
H = struct();
H.save = @savebin;
H.load = @loadbin;
H.dir  = @datadir;
end

function d = datadir()
d = fullfile(fileparts(mfilename('fullpath')),'data');
if ~exist(d,'dir'), mkdir(d); end
end

function savebin(name,A)
d = datadir();
A = double(A);
fid = fopen(fullfile(d,[name '.bin']),'wb');
fwrite(fid,A,'float64');            % column-major
fclose(fid);
fid = fopen(fullfile(d,[name '.shape']),'w');
fprintf(fid,'%d ',size(A));
fclose(fid);
end

function A = loadbin(name)
d = datadir();
fid = fopen(fullfile(d,[name '.shape']),'r');
shape = fscanf(fid,'%d');
fclose(fid);
fid = fopen(fullfile(d,[name '.bin']),'rb');
A = fread(fid,prod(shape),'float64');
fclose(fid);
shape = shape(:)';
if numel(shape) < 2, shape = [shape 1]; end   % numpy 1-D -> Octave column
A = reshape(A,shape);
end

File: Archive/deconv_pipeline/binio.py
"""Flat float64 binary exchange between Octave and Python.

Octave writes column-major, so every array is read back with order='F'.
Shapes live in a sidecar '<name>.shape' text file so neither side has to
guess.
"""
import os
import numpy as np

ROOT = os.path.dirname(os.path.abspath(__file__))
DATA = os.path.join(ROOT, 'data')
os.makedirs(DATA, exist_ok=True)


def save(name, arr):
    arr = np.asarray(arr, dtype=np.float64)
    with open(os.path.join(DATA, name + '.bin'), 'wb') as f:
        f.write(np.asfortranarray(arr).tobytes(order='F'))
    with open(os.path.join(DATA, name + '.shape'), 'w') as f:
        f.write(' '.join(str(s) for s in arr.shape))


def load(name):
    with open(os.path.join(DATA, name + '.shape')) as f:
        shape = tuple(int(s) for s in f.read().split())
    a = np.fromfile(os.path.join(DATA, name + '.bin'), dtype=np.float64)
    return a.reshape(shape, order='F')


def exists(name):
    return os.path.exists(os.path.join(DATA, name + '.bin'))
